coriolis/licensing/client.py

```diff
@@ -0,0 +1,188 @@
+# Copyright 2019 Cloudbase Solutions Srl
+# All Rights Reserved.
+
+import json
+import os
+import requests
+
+from coriolis import exception
+from coriolis import utils
+from oslo_log import log as logging
+
+
+LOG = logging.getLogger(__name__)
+
+DEFAULT_LICENSING_PROTOCOL = "http"
+DEFAULT_LICENSING_PORT = 37667
+DEFAULT_LICENSING_API_VERSION = "v1"
+
+RESERVATION_TYPE_REPLICA = "replica"
+RESERVATION_TYPE_MIGRATION = "migration"
+
+
+class LicensingClient(object):
+    """ Class for accessing the Coriolis licensing server API. """
+
+    def __init__(self, base_url, allow_untrusted=False):
+        """ :param base_url: URL for the API service, including scheme """
+        self._base_url = base_url.rstrip('/')
+        self._verify = not allow_untrusted
+
+    @classmethod
+    def from_env(cls):
+        """ Retuns a `LicensingClient` object instatiated using the
+        following env vars:
+        LICENSING_SERVER_BASE_URL="https://10.7.2.3:37667/v1"
+        LICENSING_SERVER_ALLOW_UNTRUSTED="<set to anything>"
+        Returns None if 'LICENSING_SERVER_BASE_URL' is not defined.
+        """
+        base_url = os.environ.get("LICENSING_SERVER_BASE_URL")
+        if not base_url:
+            LOG.warn(
+                "No 'LICENSING_SERVER_BASE_URL' env var present. Cannot "
+                "instantiate licensing client.")
+            return None
+        allow_untrusted = False
+        if os.environ.get("LICENSING_SERVER_ALLOW_UNTRUSTED"):
+            allow_untrusted = True
+
+        # try out client:
+        client = cls(base_url, allow_untrusted=allow_untrusted)
+        try:
+            client.get_licence_status()
+        except (Exception, KeyboardInterrupt):
+            LOG.warn(
+                "Error while testing licensing client. Exception:\n%s",
+                utils.get_exception_details())
+
+        return client
+
+    def _get_url_for_resource(self, resource):
+        """ Provides full URL for subresource.
+        Ex: "licences" -> "http://$host:$port/v1/licences"
+        """
+        return "%s/%s" % (self._base_url, resource.strip('/'))
```

Comment:
you probably want ```lstrip()``` here. Otherwise both leading and trailing slashes will be stripped.

Reply:
It was intentional, as the Go implementation of the server doesn't like trailing slashes, as the Replicator does :))

Reply:
ahh. Makes sense :).